# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/JSScript.h
# Commit: fa120d93961e
# Full Hash: fa120d93961efddd63d50135f03c55e75dac875a
# Author: Doug Thayer <dothayer@mozilla.com>
# Date: 2024-04-26 04:18:06
# Regressor Bug: 1881995
# File Overlap Count: 1
# Description:
#   Bug 1881995 - Implement ForwardedAtoms and create them during atomization r=iain,sfink
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D202690
# ==============================================================================

diff -r f24ffc487097 -r fa120d93961e js/src/vm/JSScript.h
--- a/js/src/vm/JSScript.h	Thu Apr 25 21:30:10 2024 +0300
+++ b/js/src/vm/JSScript.h	Thu Apr 25 18:29:15 2024 +0000
@@ -2040,6 +2040,23 @@
     return getString(GET_GCTHING_INDEX(pc));
   }
 
+  bool atomizeString(JSContext* cx, jsbytecode* pc) {
+    MOZ_ASSERT(containsPC<js::GCThingIndex>(pc));
+    MOZ_ASSERT(js::JOF_OPTYPE((JSOp)*pc) == JOF_STRING);
+    js::GCThingIndex index = GET_GCTHING_INDEX(pc);
+    JSString* str = getString(index);
+    if (str->isAtom()) {
+      return true;
+    }
+    JSAtom* atom = js::AtomizeString(cx, str);
+    if (!atom) {
+      return false;
+    }
+    js::gc::CellPtrPreWriteBarrier(data_->gcthings()[index]);
+    data_->gcthings()[index] = JS::GCCellPtr(atom);
+    return true;
+  }
+
   JSAtom* getAtom(js::GCThingIndex index) const {
     return &gcthings()[index].as<JSString>().asAtom();
   }